File: java/packable/src/main/java/io/packable/DecodeBuffer.java
package io.packable;

final class DecodeBuffer {
    final byte[] hb;
    int position;
    int limit;

    DecodeBuffer(byte[] array, int offset, int length) {
        hb = array;
        position = offset;
        limit = offset + length;
    }

    void checkBound(int offset, int len) {
        if (offset + len > limit) {
            throw new IndexOutOfBoundsException("buffer out of bound");
        }
    }

    public boolean hasRemaining() {
        return position < limit;
    }

    public byte readByte() {
        return hb[position++];
    }

    public short readShort() {
        return (short) (((hb[position++] & 0xff)) |
                ((hb[position++]) << 8));
    }

    public int readInt() {
        int value = (((hb[position] & 0xff)) |
                ((hb[position + 1] & 0xff) << 8) |
                ((hb[position + 2] & 0xff) << 16) |
                ((hb[position + 3]) << 24));
        position += 4;
        return value;
    }

    public int readVarInt32() {
        int x = hb[position++];
        if ((x >> 7) == 0) return x;
        x = (x & 0x7f) | (hb[position++] << 7);
        if ((x >> 14) == 0) return x;
        x = (x & 0x3fff) | (hb[position++] << 14);
        if ((x >> 21) == 0) return x;
        x = (x & 0x1fffff) | (hb[position++] << 21);
        if ((x >> 28) == 0) return x;
        x = (x & 0xfffffff) | (hb[position++] << 28);
        return x;
    }

    public long readLong(int i) {
        return (((long) hb[i] & 0xff) |
                (((long) hb[i+1] & 0xff) << 8) |
                (((long) hb[i+2] & 0xff) << 16) |
                (((long) hb[i+3] & 0xff) << 24) |
                (((long) hb[i+4] & 0xff) << 32) |
                (((long) hb[i+5] & 0xff) << 40) |
                (((long) hb[i+6] & 0xff) << 48) |
                (((long) hb[i+7]) << 56));
    }

    public long readLong() {
        long value = readLong(position);
        position += 8;
        return value;
    }

    public float readFloat() {
        return Float.intBitsToFloat(readInt());
    }

    public double readDouble() {
        return Double.longBitsToDouble(readLong());
    }

    public void readBytes(byte[] bytes) {
        int len = bytes.length;
        System.arraycopy(hb, position, bytes, 0, len);
        position += len;
    }
}

File: kmp-packable/composeApp/src/commonMain/kotlin/io/packable/DecodeBuffer.kt
package io.packable

internal class DecodeBuffer(
    val hb: ByteArray,
    var position: Int,
    length: Int
) {
    var limit: Int = position + length

    fun checkBound(offset: Int, len: Int) {
        if (offset + len > limit) {
            throw IndexOutOfBoundsException("buffer out of bound")
        }
    }

    fun hasRemaining(): Boolean {
        return position < limit
    }

    fun readByte(): Byte {
        return hb[position++]
    }

    fun readShort(): Short {
        return (((hb[position++].toInt() and 0xff)) or
                ((hb[position++]).toInt() shl 8)).toShort()
    }

    fun readInt(): Int {
        return ((((hb[position++].toInt() and 0xff)) or
                ((hb[position++].toInt() and 0xff) shl 8) or
                ((hb[position++].toInt() and 0xff) shl 16) or
                ((hb[position++]).toInt() shl 24)))
    }

    fun readVarInt32(): Int {
        var x: Int = hb[position++].toInt()
        if ((x shr 7) == 0) return x
        x = (x and 0x7f) or (hb[position++].toInt() shl 7)
        if ((x shr 14) == 0) return x
        x = (x and 0x3fff) or (hb[position++].toInt() shl 14)
        if ((x shr 21) == 0) return x
        x = (x and 0x1fffff) or (hb[position++].toInt() shl 21)
        if ((x shr 28) == 0) return x
        x = (x and 0xfffffff) or (hb[position++].toInt() shl 28)
        return x
    }

    fun readLong(index: Int): Long {
        var i: Int = index
        return (((hb[i++].toLong() and 0xffL) or
                ((hb[i++].toLong() and 0xffL) shl 8) or
                ((hb[i++].toLong() and 0xffL) shl 16) or
                ((hb[i++].toLong() and 0xffL) shl 24) or
                ((hb[i++].toLong() and 0xffL) shl 32) or
                ((hb[i++].toLong() and 0xffL) shl 40) or
                ((hb[i++].toLong() and 0xffL) shl 48) or
                ((hb[i].toLong()) shl 56)))
    }

    fun readLong(): Long {
        val value: Long = readLong(position)
        position += 8
        return value
    }

    fun readFloat(): Float {
        return Float.fromBits(readInt())
    }

    fun readDouble(): Double {
        return Double.fromBits(readLong())
    }

    fun readBytes(bytes: ByteArray) {
        val len: Int = bytes.size
        hb.copyInto(bytes, 0, position, position + len)
        position += len
    }
}
